### PR TITLE
Bugfix: avoid clobbering of valid variants when a config key contains a superset of values of the same key in a lower priority cbc.yaml

### PR DESCRIPTION
**Pull request by danpetry**
### Description

Fixes https://github.com/conda/conda-build/issues/5038

When higher priority cbc.yamls contain, for a certain config entry, a superset of values that are in a lower priority cbc.yaml, the extra values get ignored. For example, if you have in a feedstock's conda_build_config.yaml

```
openssl:
  - 1.1.1
  - 3.0
```

and in aggregate/conda_build_config.yaml

```
openssl:
  - 3.0
```

I would expect two versions of the package is build, but only the one with openssl 3.0 is built.

When a `zip_keys` entry is used in the feedstock's cbc.yaml, the keys get zipped, but then the variants without the entry from the aggregate cbc.yaml gets clobbered.

### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [x] Add / update necessary tests?
- [ ] Add / update outdated documentation?

### CODE REVIEW

**Comment by travishathaway** [comment]
@conda-bot check

**Comment by danpetry** [comment]
Hi @travishathaway or @kenodegard or other? It would be nice to get some eyes on this if possible - we're affected by this relatively frequently in the packaging team

File: conda_build/variants.py [hunk 1]
```diff
@@ -679,16 +679,21 @@ def filter_combined_spec_to_used_keys(combined_spec, specs):
 
     # TODO: act here?
     combined_spec = explode_variants(combined_spec)
+    seen_keys = (
+        set()
+    )  # seen_keys makes sure that a setting from a lower-priority spec doesn't clobber the same setting that has been redefined in a higher-priority spec.
```
[the remark this answers is not in the document]
Comment (danpetry) [reply conda_build/variants.py:684]: agreed, the pre-commit checks told me to do this though.

File: tests/test_variants.py [hunk 2]
```diff
@@ -657,3 +658,45 @@ def test_variant_subkeys_retained():
...
+
+    combined_spec = low_prio_config.copy()
+    combined_spec.update(high_prio_config)
+    specs = OrderedDict(
+        [
+            ("internal_defaults", internal_defaults),
+            ("low_prio_config", low_prio_config),
+            ("high_prio_config", high_prio_config),
+        ]
+    )
```
Comment (kenodegard) [inline tests/test_variants.py:700]: no need to use `OrderedDict` anymore:

```suggestion
    combined_spec = {
        **low_prio_config,
        **high_prio_config,
    }
    specs = {
        "internal_defaults": internal_defaults,
        "low_prio_config": low_prio_config,
        "high_prio_config": high_prio_config,
    }
```

File: tests/test_variants.py [hunk 3]
```diff
@@ -657,3 +658,45 @@ def test_variant_subkeys_retained():
...
+
+@pytest.mark.parametrize(
+    "internal_defaults, low_prio_config, high_prio_config, expected",
+    [
+        (
+            {"pkg_1": "1.0"},
+            {"pkg_1": "1.1"},
+            {"pkg_1": ["1.1", "1.2"], "pkg_2": ["1.1"]},
+            [{"pkg_1": "1.1", "pkg_2": "1.1"}, {"pkg_1": "1.2", "pkg_2": "1.1"}],
+        ),
```
Comment (kenodegard) [inline tests/test_variants.py:671]:
```suggestion
        pytest.param(
            {"pkg_1": "1.0"},
            {"pkg_1": "1.1"},
            {"pkg_1": ["1.1", "1.2"], "pkg_2": ["1.1"]},
            [{"pkg_1": "1.1", "pkg_2": "1.1"}, {"pkg_1": "1.2", "pkg_2": "1.1"}],
            id="basic",
        ),
```

File: tests/test_variants.py [hunk 4]
```diff
@@ -657,3 +658,45 @@ def test_variant_subkeys_retained():
...
+            {
+                "pkg_1": ["1.1", "1.2"],
+                "pkg_2": ["1.1", "1.2"],
+                "zip_keys": [["pkg_1", "pkg_2"]],
+            },
+            [
+                {"pkg_1": "1.1", "pkg_2": "1.1", "zip_keys": [["pkg_1", "pkg_2"]]},
+                {"pkg_1": "1.2", "pkg_2": "1.2", "zip_keys": [["pkg_1", "pkg_2"]]},
+            ],
+        ),
```
Comment (kenodegard) [inline tests/test_variants.py:684]:
```suggestion
        pytest.param(
            {"pkg_1": "1.0"},
            {"pkg_1": "1.1"},
            {
                "pkg_1": ["1.1", "1.2"],
                "pkg_2": ["1.1", "1.2"],
                "zip_keys": [["pkg_1", "pkg_2"]],
            },
            [
                {"pkg_1": "1.1", "pkg_2": "1.1", "zip_keys": [["pkg_1", "pkg_2"]]},
                {"pkg_1": "1.2", "pkg_2": "1.2", "zip_keys": [["pkg_1", "pkg_2"]]},
            ],
            id="zip_keys",
        ),
```

**Review by kenodegard — CHANGES_REQUESTED** [review]

**Review by kenodegard — APPROVED** [review]

**Comment by danpetry** [comment]
Thanks @kenodegard ! how many approvals required to merge?

**Pull request merged**